subprojects/gradle/process/src/main/kotlin/com/avito/utils/ProcessRunner.kt

```diff
@@ -1,129 +1,147 @@
         }
-        .start()
-}
-
-@JvmOverloads
-fun runCommand(
-    command: String,
-    workingDirectory: File? = null,
-    loggerFactory: LoggerFactory
-): Result<String> = try {
-    val binary = splitCommand(command).firstOrNull()
-
-    val logger = loggerFactory.create(binary ?: "runCommand")
-
-    val process = spawnProcess(
-        command = command,
-        workingDirectory = workingDirectory
-    )
-
-    val processOutput = StringBuffer()
-
-    val outputThread = IOThread(
-        source = process.inputStream,
-        result = processOutput,
-        logger = logger
-    )
 
-    val errorThread = IOThread(
-        source = process.errorStream,
-        result = processOutput,
-        logger = logger
-    )
+        override fun run(
+            command: String,
+            timeout: Long,
+            unit: TimeUnit
+        ): Result<String> {
+            return runCommand(
+                command = command,
+                workingDirectory = workingDirectory,
+                timeout = Timeout(timeout, unit)
+            )
+        }
 
-    outputThread.start()
-    errorThread.start()
+        override fun spawn(command: String, outputTo: File?): Process {
+            return spawnProcess(command, workingDirectory, outputTo)
+        }
 
-    outputThread.join()
-    errorThread.join()
-    process.waitFor()
+        /**
+         * [ProcessBuilder] expects command as an arguments list, arguments could be separated by space
+         */
+        @VisibleForTesting
+        fun splitCommand(source: String): Array<String> = Commandline.translateCommandline(source)
+
+        private fun spawnProcess(
+            command: String,
+            workingDirectory: File? = null,
+            outputTo: File? = null
+        ): Process {
+            return ProcessBuilder(*splitCommand(command))
+                .also { builder ->
+                    if (workingDirectory != null) {
+                        builder.directory(workingDirectory)
+                    }
+                    if (outputTo != null) {
+                        builder.redirectOutput(outputTo)
+                    }
+                }
+                .start()
+        }
 
-    val output = processOutput.toString().trim()
+        private class Timeout(val time: Long, val unit: TimeUnit)
+
+        private fun runCommand(
+            command: String,
+            workingDirectory: File? = null,
+            timeout: Timeout?
+        ): Result<String> = try {
+            val process = spawnProcess(
+                command = command,
+                workingDirectory = workingDirectory
+            )
+            val outputFuture = cliOutputReaderThreadPool.submit(
+                CLIOutputReaderTask(
+                    stdout = process.inputStream,
+                    stderr = process.errorStream
+                )
+            )
+            val output = if (timeout != null) {
+                process.waitFor(timeout.time, timeout.unit)
```

Comment:
before we could hang and that blocks thread from what we invoke `process run` for example from KubernetesReservationProvider when initializing device

Now we could configure the timeout for CLI command if need